### PR TITLE
fix: make user creation errors helpful

### PR DESCRIPTION
**Pull request by eliobischof**
- Returns more differentiated error messages on failed user validations
- Adds missing translations
- Uses type aliases PhoneNumber and EmailAddress so validation and normalization is explicitly handled centrally in the domain package.

Complements #5409 
Follow-up: #5412 
Closes: #5038 

```[tasklist]
### Definition of Ready
- [x] I am happy with the code
- [x] Short description of the feature/issue is added in the pr description
- [x] PR is linked to the corresponding user story
- [x] Acceptance criteria are met
- [x] All open todos and follow ups are defined in a new ticket and justified
- [x] Deviations from the acceptance criteria and design are agreed with the PO and documented.
- [x] No debug or dead code
- [x] Critical parts are tested automatically
- [x] Where possible E2E tests are implemented
- [x] Documentation/examples are up-to-date
- [x] All non-functional requirements are met
- [ ] Functionality of the acceptance criteria is checked manually on the dev system.
```

### CODE REVIEW

File: e2e/config/host.docker.internal/zitadel.yaml [hunk 1]
```diff
@@ -3,7 +3,7 @@ ExternalSecure: false
 
 Database:
   cockroach:
-    Host:  db
+    Host:  host.docker.internal
```
Comment (eliobischof) [inline e2e/config/host.docker.internal/zitadel.yaml:6]: This makes the config reusable with an out-of-docker zitadel process and an /etc/hosts entry

File: e2e/config/localhost/zitadel.yaml [hunk 2]
```diff
@@ -3,7 +3,7 @@ ExternalSecure: false
 
 Database:
   cockroach:
-    Host:  db
+    Host: host.docker.internal
```
Comment (eliobischof) [inline e2e/config/localhost/zitadel.yaml:6]: This makes the e2e config reusable with an out-of-docker zitadel

File: internal/query/projection/user.go [hunk 3]
```diff
@@ -342,7 +342,7 @@ func (p *userProjection) reduceHumanAdded(event eventstore.Event) (*handler.Stat
 				handler.NewCol(HumanPreferredLanguageCol, &sql.NullString{String: e.PreferredLanguage.String(), Valid: !e.PreferredLanguage.IsRoot()}),
 				handler.NewCol(HumanGenderCol, &sql.NullInt16{Int16: int16(e.Gender), Valid: e.Gender.Specified()}),
 				handler.NewCol(HumanEmailCol, e.EmailAddress),
-				handler.NewCol(HumanPhoneCol, &sql.NullString{String: e.PhoneNumber, Valid: e.PhoneNumber != ""}),
+				handler.NewCol(HumanPhoneCol, &sql.NullString{String: string(e.PhoneNumber), Valid: e.PhoneNumber != "" /* TODO: Really validate */}),
```
Comment (eliobischof) [inline internal/query/projection/user.go:345]: @adlerhurst do we want that?

File: internal/domain/human_phone.go [hunk 4]
```diff
@@ -27,17 +39,16 @@ type PhoneCode struct {
...
+	if p == nil {
+		return caos_errs.ThrowInvalidArgument(nil, "PHONE-YlbwO", "Errors.User.Phone.Empty")
+	}
+	normalizedNumber, err := p.PhoneNumber.Normalize()
 	if err != nil {
-		return caos_errs.ThrowInvalidArgument(nil, "EVENT-so0wa", "Errors.User.Phone.Invalid")
+		return err
 	}
-	p.PhoneNumber = libphonenumber.Format(phoneNr, libphonenumber.E164)
+	// TODO: Let's stop mutating state
```
Comment (eliobischof) [inline internal/domain/human_phone.go:50]: Follow up issue created: https://github.com/zitadel/zitadel/issues/5412

Comment (adlerhurst) [reply internal/domain/human_phone.go:50]: Please add the comment to the file directly

File: e2e/config/host.docker.internal/zitadel.yaml (hunk 1 above)
Comment (adlerhurst) [reply e2e/config/host.docker.internal/zitadel.yaml:6]: Please add the comment to the file

File: internal/domain/human.go [hunk 5]
```diff
@@ -60,16 +61,30 @@ func (f Gender) Specified() bool {
 	return f > GenderUnspecified && f < genderCount
 }
 
-func (u *Human) IsValid() bool {
-	return u.Username != "" && u.Profile != nil && u.Profile.IsValid() && u.Email != nil && u.Email.IsValid() && u.Phone == nil || (u.Phone != nil && u.Phone.PhoneNumber != "" && u.Phone.IsValid())
+func (u *Human) Normalize() error {
```
Comment (stebenz) [inline internal/domain/human.go:64]: No differentiation between normalization and validation?

File: internal/domain/human_email.go [hunk 6]
```diff
@@ -2,20 +2,38 @@ package domain
...
+	if e == "" {
+		return errors.ThrowInvalidArgument(nil, "EMAIL-spblu", "Errors.User.Email.Empty")
+	}
+	if !emailRegex.MatchString(string(e)) {
+		return errors.ThrowInvalidArgument(nil, "EMAIL-599BI", "Errors.User.Email.Invalid")
+	}
+	return nil
+}
+
+func (e EmailAddress) Normalize() EmailAddress {
```
Comment (stebenz) [inline internal/domain/human_email.go:29]: Why does Normalize handle differently for email and phone with the return values?

File: internal/user/model/email.go [hunk 7]
```diff
@@ -21,10 +21,6 @@ type EmailCode struct {
 	Expiry time.Duration
 }
 
-func (e *Email) IsValid() bool {
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply internal/user/model/email.go:24]: Which old views? The tests compile and the binary compiles, so it isn't used, right?

File: internal/domain/human.go (hunk 5 above)
Comment (eliobischof) [reply internal/domain/human.go:64]: To make it explicit that state is changed, I renamed the method. But I didn't want to change existing logic where the effects are not obvious.
Instead I created a follow-up issue that addresses this https://github.com/zitadel/zitadel/pull/5382#discussion_r1131261119

File: internal/domain/human_email.go (hunk 6 above)
Comment (eliobischof) [reply internal/domain/human_email.go:29]: see above

File: internal/user/model/phone.go [hunk 8]
```diff
@@ -27,20 +25,6 @@ type PhoneCode struct {
 	Expiry time.Duration
 }
 
-func (p *Phone) IsValid() bool {
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply internal/user/model/phone.go:30]: see above

File: internal/domain/human_phone.go (hunk 4 above)
Comment (eliobischof) [reply internal/domain/human_phone.go]: done

File: internal/query/projection/user.go [hunk 9]
```diff
@@ -399,7 +399,7 @@ func (p *userProjection) reduceHumanRegistered(event eventstore.Event) (*handler
 				handler.NewCol(NotifyUserIDCol, e.Aggregate().ID),
 				handler.NewCol(NotifyInstanceIDCol, e.Aggregate().InstanceID),
 				handler.NewCol(NotifyLastEmailCol, e.EmailAddress),
-				handler.NewCol(NotifyLastPhoneCol, &sql.NullString{String: e.PhoneNumber, Valid: e.PhoneNumber != ""}),
+				handler.NewCol(NotifyLastPhoneCol, &sql.NullString{String: string(e.PhoneNumber), Valid: e.PhoneNumber != "" /* TODO: Really validate */}),
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply internal/query/projection/user.go]: done

File: internal/query/projection/user.go [hunk 10]
```diff
@@ -660,7 +660,7 @@ func (p *userProjection) reduceHumanPhoneChanged(event eventstore.Event) (*handl
 		),
 		crdb.AddUpdateStatement(
 			[]handler.Column{
-				handler.NewCol(NotifyLastPhoneCol, &sql.NullString{String: e.PhoneNumber, Valid: e.PhoneNumber != ""}),
+				handler.NewCol(NotifyLastPhoneCol, &sql.NullString{String: string(e.PhoneNumber), Valid: e.PhoneNumber != "" /* TODO: Really validate */}),
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply internal/query/projection/user.go]: done

File: internal/query/projection/user.go [hunk 11]
```diff
@@ -786,7 +786,7 @@ func (p *userProjection) reduceHumanEmailChanged(event eventstore.Event) (*handl
 		),
 		crdb.AddUpdateStatement(
 			[]handler.Column{
-				handler.NewCol(NotifyLastEmailCol, &sql.NullString{String: e.EmailAddress, Valid: e.EmailAddress != ""}),
+				handler.NewCol(NotifyLastEmailCol, &sql.NullString{String: string(e.EmailAddress), Valid: e.EmailAddress != "" /* TODO: e.EmailAddress.Validate() == nil */}),
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply internal/query/projection/user.go]: done

File: e2e/config/host.docker.internal/zitadel.yaml (hunk 1 above)
Comment (eliobischof) [reply e2e/config/host.docker.internal/zitadel.yaml:7]: done

File: internal/query/projection/user.go (hunk 3 above)
Comment (eliobischof) [reply internal/query/projection/user.go]: done

File: internal/query/projection/user.go [hunk 12]
```diff
@@ -390,7 +390,7 @@ func (p *userProjection) reduceHumanRegistered(event eventstore.Event) (*handler
 				handler.NewCol(HumanPreferredLanguageCol, &sql.NullString{String: e.PreferredLanguage.String(), Valid: !e.PreferredLanguage.IsRoot()}),
 				handler.NewCol(HumanGenderCol, &sql.NullInt16{Int16: int16(e.Gender), Valid: e.Gender.Specified()}),
 				handler.NewCol(HumanEmailCol, e.EmailAddress),
-				handler.NewCol(HumanPhoneCol, &sql.NullString{String: e.PhoneNumber, Valid: e.PhoneNumber != ""}),
+				handler.NewCol(HumanPhoneCol, &sql.NullString{String: string(e.PhoneNumber), Valid: e.PhoneNumber != "" /* TODO: Really validate */}),
```
[the remark this answers is not in the document]
Comment (eliobischof) [reply internal/query/projection/user.go]: done

File: internal/user/model/email.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/user/model/email.go:24]: Shouldn't yes

File: internal/user/model/phone.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/user/model/phone.go:30]: Shouldn't yes

File: internal/domain/human.go (hunk 5 above)
Comment (stebenz) [reply internal/domain/human.go:64]: What I meant is that the Normalize functions for the different structs handle differently, which shouldn't be the case right?

Comment (eliobischof) [reply internal/domain/human.go:64]: It shouldn't, thats why there is a follow up issue. It is already wrong on main.

**Review by stebenz — APPROVED** [review]

**Pull request merged**